src/utils/tracked-controls.js

```diff
@@ -78,46 +72,46 @@ module.exports.isControllerPresent = isControllerPresent;
- *   If filterHand is not set, we skip until we have found the nth matching controller, where n equals targetControllerNumber
+ *     is unhanded (controller.hand is ''), we skip until we have found a number of matching controllers that equals filterControllerIndex
+ *   If filterHand is not set, we skip until we have found the nth matching controller, where n equals filterControllerIndex
  *
- * The method should be called with one of: [filterIdExact, filterIdPrefix] AND one or both of: [filterHand, targetControllerNumber]
+ * The method should be called with one of: [filterIdExact, filterIdPrefix] AND one or both of: [filterHand, filterControllerIndex]
  *
  * @param {object} controllers - Array of gamepads to search
  * @param {string} filterIdExact - If set, used to find controllers with id === this value
  * @param {string} filterIdPrefix - If set, used to find controllers with id startsWith this value
  * @param {object} filterHand - If set, further filters controllers with matching 'hand' property
- * @param {object} targetControllerNumber - Find the nth matching controller, where n equals targetControllerNumber. defaults to 0.
+ * @param {object} filterControllerIndex - Find the nth matching controller, where n equals filterControllerIndex. defaults to 0.
  */
-function findMatchingController (controllers, filterIdExact, filterIdPrefix, filterHand, targetControllerNumber) {
+function findMatchingController (controllers, filterIdExact, filterIdPrefix, filterHand, filterControllerIndex) {
   var controller;
   var i;
   var matchingControllerOccurence = 0;
-  targetControllerNumber = targetControllerNumber || 0;
+  var targetControllerMatch = filterControllerIndex || 0;
 
   for (i = 0; i < controllers.length; i++) {
     controller = controllers[i];
     // Determine if the controller ID matches our criteria
-    if (filterIdPrefix && controller.id.indexOf(filterIdPrefix) === -1) continue;
-    if (!filterIdPrefix && controller.id !== filterIdExact) continue;
-
-    if (filterHand) {
-      if (filterHand === controller.hand) {
-        // If the component requests a specific hand and found a matching one, we ignore the
-        // targetControllerNumber requirement and early exit.
-        return controller;
-      } else if (controller.hand) {
-        continue;
-      }
-      // If we reach here, the controller is unhanded - check against targetControllerNumber
+    if (filterIdPrefix && controller.id.indexOf(filterIdPrefix) === -1) { continue; }
+    if (!filterIdPrefix && controller.id !== filterIdExact) { continue; }
+
+    // If the hand filter and controller handedness are defined we compare them.
+    if (filterHand && controller.hand && filterHand !== controller.hand) { continue; }
+
+    // If we have detected an unhanded controller and the component was asking for a particular hand,
+    // we need to treat the controllers in the array as pairs of controllers. This effectively means that we
```

Comment:
I am not sure this is needed or desirable; originally there were issues with browsers not uniformly exposing handedness, but that is gone now, so if the filter is requesting a hand, we should require one.

Still need coverage for the case where we want to match no-handedness, my suggestion is to use the string `none` such as the following:
```
if (filterHand && (filterHand !== (controller.hand ? controller.hand : 'none'))) { continue; }
```

Reply:
Why do we need `none`? In the absence of a filterHand the condition is a pass through.

Reply:
For example - trackers do not have handedness.  Also, if you want to use a third vive controller as spectator camera, my understanding is that the first two get hands, then the rest are no-handed (hand undefined) -- maybe that has changed.

Reply:
(This bit is inside tracked-controls utils that gets used for everything with a pose, not just within the windows-motion-controls component)

Reply:
Let's deal with the problem at hand and we will accommodate those no handedness / 3rd control edge cases when we integrate the trackers.

Reply:
right, as I said, I'm fine with that, this comment is really more about the flagged code portion:
```
// If we have detected an unhanded controller and the component was asking for a particular hand,
// we need to treat the controllers in the array as pairs of controllers.
```
I don't think we need to do that, nor should we.
If filterHand is supplied, it should be used.  e.g. hand left, index 1 (to get the second left hand controller, instead of 0 for the first)

Reply:
Yeah I agree this condition is a bit cryptic. The case is trying to cover is if there are more than 2 controllers connected the  filterControllerIndex will control which pair in the list is selected as left / right. I do not think this is a possible scenario on any platform today. I'm happy to let `left` === 0 and `right` === 1 if the `controller.hand` is not supplied.

Reply:
These kind of details taken care for you is one of the reasons why people should be using A-Frame 😄 All those edge cases are not fun to deal with.

Reply:
But that is not correct behavior.  Unlike Vive controllers, the Windows motion controllers are handed, more like Oculus Touch.  So handedness is not a function of cardinality

Reply:
developer.microsoft.com/en-us/windows/mixed-reality/motion_controllers

Look closely at the picture.

Reply:
(Microsoft Edge does in fact provide the hand property for appropriate gamepads/controllers.)

Reply:
It's ok. If there's always a hand the filterHand is always supplied by `windows-motion-controls` so you won't reach this statement https://github.com/aframevr/aframe/pull/3013/files/638c6a2a2fad503ba67f291bc8a1d7b8df3e05d5#diff-bc91e5db13c01778a4dfb59ce1fd3f05R27

Reply:
I seem to recall that the original mapping for Vive, before browsers exposed hand, was actually the opposite... 0 right, 1 left.  Anyway, now that browsers do the right thing, if the Windows motion controls behave that way, I'm not sure this index mapping should be applied more generally in tracked-controls used by all, versus in the windows-motion-controls component... since the convention may apparently vary by controller type.

Reply:
If the filterHand is always supplied, why do we need this code snippet at all?  Now I am confused...

Reply:
Meaning, the way the current code is designed, hand applies before index, so you ask for the second left handed controller with filter of hand=left and index=1 (since they start at zero).  No need to worry about whether pairs or not.

Reply:
Vive handedness is determined at runtime based on the relative position to the head. I believe there's still a chance of `controller.hand` to come undefined.

Reply:
Right, but if you specified that you wanted a handed controller, shouldn't you only get it when the system agrees that it is left hand?  Otherwise you will end up with hands changing abruptly for no reason

Reply:
Thanks for indulging the long discussion.  My sense is that the windows-motion-controls PR can handle left hand, right hand, and multiple pairs without the changes to tracked-controls utils (which have been tested to work properly across Vive, Oculus touch, daydream and gearvr).  If we want to change behavior due to Vive indeterminate hand, I suggest we do that in separate PR

Reply:
Now I am starting to get confused too!

If, as you say, all browsers are now exposing the hand property correctly - it is very tempting to update the vive controls class to use that. We do still need to consider the 'unhanded case' though as 
the underlying windows API that drives the Motion Controllers support in Edge's WebVR does support unhanded controllers, even though there aren't any on the market at this time. 
(The current Motion Controllers are, as you observed, handed). If our platform exposes an unhanded controller, the gamepad 'hand' property will be an empty string: ''.

The reason I have the (yes, cryptic) snippet is that the underlying platform also supports more than 2 unhanded controls connected at once. So in theory the platform could tell Edge that the user has 4 or more unhanded controllers connected. The question is - do we want these to work with the hand-controls component? If so, we need them to work with a component that has requested a particular hand.

If so, we need to keep the logic here that puts unhanded controllers into pairs of left and right, based on their index.

Regarding the number of connected handed controllers, it is easy to connect 4 handed controllers today using two physical, and two simulated via the windows mixed reality simulator.

Reply:
@machenmusik Do you think the code we have here is reasonable even tough there might be an edge case or two we're missing? Can we go with this and refine when integrating the Vive trackers and test on different browsers if our assumptions about `handeness` still apply?

Reply:
@leweaver When doing #2583 I did in fact switch over vive-controls to use handedness, no more weird index stuff.

@dmarcos When doing #2583, I had tested against all supported controllers as well as vive-trackers, which is why I am personally more comfortable with the changes to tracked-controls utils there.  (Although it has been so long that other changes may have crept in that break it now, I suppose.)

On unhanded controllers - it sounds as though there will be WinMR equivalents of vive-trackers, in which case the main changes in #2583 may prove useful as well, namely handling no-handedness (although it sounds as though we now have empty string to contend with as well as undefined) and also adding full rotation offset to tracked-controls, not just Z (because the native orientation of vive-tracker is somewhat different from canonical model).

I am happy to welcome @leweaver to join the A-Frame project and thus take over WinMR controller support, just trying to share learnings from doing the last few.  Feel free to take whatever is worthwhile from #2583 and the spatial-controls branch I had done, if it is helpful.

As for whether unhandled controllers should be allowed to match handed requests, that's a bit of a philosophical question.  The dynamic Vive wand case is simpler, as the controllers will in all likelihood gain handedness eventually.  In the WinMR and Vive tracker case, it's less clear, although I would expect unhandled WinMR controllers to have different id strings and thus be handled by a related but separate component.

Reply:
Ok, let's keep this part of the PR as is and improve it later in follow up PRs

Reply:
> If so, we need to keep the logic here that puts unhanded controllers into pairs of left and right, based on their index.

The problem I see with this is that all unhandled controllers may not be created equal.  Let's assume that unhandled controllers include totems that can be used in either hand (with no dynamic hand assignment), and also other things like gun controllers as are popular with PSVR.  The user has limited or no control over enumerated order.  What if the enumerated order is totem, gun, totem?  Then you have two left-handed totems and a gun for a right hand, which may not be the intended outcome...

Reply:
> If so, we need to keep the logic here that puts unhanded controllers into pairs of left and right, based on their index.

The problem I see with this is that all unhandled controllers may not be created equal.  Let's assume that unhandled controllers include totems that can be used in either hand (with no dynamic hand assignment), and also other things like gun controllers as are popular with PSVR.  The user has limited or no control over enumerated order.  What if the enumerated order is totem, gun, totem?  Then you have two left-handed totems and a gun for a right hand, which may not be the intended outcome...

Reply:
> If so, we need to keep the logic here that puts unhanded controllers into pairs of left and right, based on their index.

The problem I see with this is that all unhandled controllers may not be created equal.  Let's assume that unhandled controllers include totems that can be used in either hand (with no dynamic hand assignment), and also other things like gun controllers as are popular with PSVR.  The user has limited or no control over enumerated order.  What if the enumerated order is totem, gun, totem?  Then you have two left-handed totems and a gun for a right hand, which may not be the intended outcome...